server/auth/types/saml/saml_auth.ts

```diff
@@ -49,7 +49,9 @@ export class SamlAuthentication extends AuthenticationType {
   }
 
   private generateNextUrl(request: KibanaRequest): string {
-    const path = request.url.path || `${this.coreSetup.http.basePath.serverBasePath}/app/kibana`;
+    const path = request.url.path
+      ? `${this.coreSetup.http.basePath.serverBasePath}${request.url.path}`
+      : `${this.coreSetup.http.basePath.serverBasePath}/app/kibana`;
```

Comment:
`` `${this.coreSetup.http.basePath.serverBasePath}${request.url.path || '/app/kibana'}` `` should works. 
Or you can also do `this.coreSetup.http.basePath.serverBasePath + (request.url.path || '/app/kibana')`